packages/cdktf/lib/app.ts

```diff
@@ -109,3 +148,59 @@ export class App extends Construct {
     }
   }
 }
+
+/**
+ * Invoke aspects on the given construct tree.
+ *
+ * originally from https://github.com/aws/aws-cdk/blob/dcae3eead0dbf9acb1ed80ba95bb104c64cb1bd7/packages/@aws-cdk/core/lib/private/synthesis.ts#L99-L137
+ */
+function invokeAspects(root: IConstruct) {
+  const invokedByPath: { [nodePath: string]: IAspect[] } = {};
+
+  let nestedAspectWarning = false;
+  recurse(root, []);
+
+  function recurse(construct: IConstruct, inheritedAspects: IAspect[]) {
+    const node = construct.node;
+    const aspects = Aspects.of(construct);
+    const allAspectsHere = [...(inheritedAspects ?? []), ...aspects.all];
+    const nodeAspectsCount = aspects.all.length;
+    for (const aspect of allAspectsHere) {
+      let invoked = invokedByPath[node.path];
+      if (!invoked) {
+        invoked = invokedByPath[node.path] = [];
+      }
+
+      if (invoked.includes(aspect)) {
+        continue;
+      }
+
+      aspect.visit(construct);
+
+      // if an aspect was added to the node while invoking another aspect it will not be invoked, emit a warning
+      // the `nestedAspectWarning` flag is used to prevent the warning from being emitted for every child
+      if (!nestedAspectWarning && nodeAspectsCount !== aspects.all.length) {
+        Annotations.of(construct).addWarning(
+          "We detected an Aspect was added via another Aspect, and will not be applied"
+        );
+        nestedAspectWarning = true;
+      }
+
+      // mark as invoked for this node
+      invoked.push(aspect);
+    }
+
+    for (const child of construct.node.children) {
+      recurse(child, allAspectsHere);
+    }
```

Comment:
I believe it's the same behaviour as https://github.com/aws/constructs/blob/ea0522e9a40ff013751734b9ae986786a9209b43/src/construct.ts#L460 - can you confirm?

Reply:
That's a tough question to answer 😄
It mostly should be (as in calling the Aspect on all children of the node it is registered on), but I suspect the new version is a bit more robust in not calling the same Aspect instance twice on the same resource if it has been added to a parent of said resource as well.

Reply:
Right - was mostly asking about the order of applying https://github.com/aws/constructs/blob/ea0522e9a40ff013751734b9ae986786a9209b43/src/construct.ts#L667-L678

But seems to make sense :)